Infra: Link status and type headers to descriptions

hugovk opened the pull request:
<!--

Please include the PEP number in the pull request title, example:

PEP NNN: Summary of the changes made

In addition, please sign the CLA.

For more information, please read our Contributing Guidelines (CONTRIBUTING.rst)

-->

#  Link status and type headers to descriptions

This helps people look up what the PEP status and type really mean by linking to their descriptions in PEP 1.

## Before

<img width="719" alt="image" src="https://user-images.githubusercontent.com/1324225/197347214-0c0f7a2a-0a5d-4540-9884-7fb735e3852b.png">

https://peps.python.org/pep-0700/

## After

<img width="718" alt="image" src="https://user-images.githubusercontent.com/1324225/197347244-ee688d43-8a44-4899-a59c-c165121dacde.png">

https://pep-previews--2842.org.readthedocs.build/pep-0700/

# Link PEP 0 type/status keys to more info in PEP 1

And on PEP 0, we have a list of types and statuses, let's also link these to the same places in PEP 1.

Plus turn them from quoted text to bullet points.

## Before

<img width="421" alt="image" src="https://user-images.githubusercontent.com/1324225/197347398-6030e418-4a46-4ddf-abbb-543c76e5a55b.png">

https://peps.python.org/pep-0000/#pep-types-key

## After

<img width="415" alt="image" src="https://user-images.githubusercontent.com/1324225/197347364-c16316a4-e1c9-4374-9466-baa6eb7e1beb.png">

https://pep-previews--2842.org.readthedocs.build/#pep-types-key

## Question

I'm not too sure about using the "§" symbol in "PEP 1 § PEP Types", but it's in the tooltip. Perhaps "PEP 1 > PEP Types" or "PEP Types" is clearer?

AA-Turner commented:
I like the PEP 0 changes, but I'm more on the fence about the PEP header ones (-0 I guess?) - everything being a link can be information overload, hiding e.g. the discussion thread.

A

CAM-Gerlach commented:
> I'm not too sure about using the "§" symbol in "PEP 1 § PEP Types", but it's in the tooltip. Perhaps "PEP 1 > PEP Types" or "PEP Types" is clearer?

I'm not sure the section name is needed at all, since you're linking directly to it. You could just say "More info in [PEP 1](LINK)" or to clarify both links aren't the same and just to the top level of PEP 1, "See [the descriptions in PEP 1](LINK)". Or even just "[More info](LINK)" for both.

> everything being a link can be information overload, hiding e.g. the discussion thread.

It seems pretty useful to have a description of the status, but I can also see the concern about making it harder for the user to see the discussion thread link and creating more semantic noise. What about instead making these `:abbr:`s and having a short summary of what the status means, e.g. Draft -> "Proposal under discussion"? This is more lightweight and avoids the noise and semantic conflation of links, while providing users more what they are likely looking for, a quick summary at their fingertips rather than a detailed formal document they have to click through and read.

hugovk commented:
> > I'm not too sure about using the "§" symbol in "PEP 1 § PEP Types", but it's in the tooltip. Perhaps "PEP 1 > PEP Types" or "PEP Types" is clearer?
> 
> I'm not sure the section name is needed at all, since you're linking directly to it. You could just say "More info in [PEP 1](LINK)" or to clarify both links aren't the same and just to the top level of PEP 1, "See [the descriptions in PEP 1](LINK)". Or even just "[More info](LINK)" for both.

Good idea, updated to "More info in [PEP 1](LINK)".

hugovk commented:
> > everything being a link can be information overload, hiding e.g. the discussion thread.
> 
> It seems pretty useful to have a description of the status, but I can also see the concern about making it harder for the user to see the discussion thread link and creating more semantic noise. What about instead making these `:abbr:`s and having a short summary of what the status means, e.g. Draft -> "Proposal under discussion"? This is more lightweight and avoids the noise and semantic conflation of links, while providing users more what they are likely looking for, a quick summary at their fingertips rather than a detailed formal document they have to click through and read.

I think `:abbr:` would be too short to get anything useful in. Shall we just drop these links?

CAM-Gerlach commented:
> Good idea, updated to "More info in [PEP 1](https://github.com/python/peps/pull/LINK)".

You might consider linking the whole phrase to the subsection ("[More info in PEP 1](https://example.com)"), to make clear that the link target is not just PEP 1 itself, but the specific information inside PEP 1. But that's a nitpick.

> I think :abbr: would be too short to get anything useful in. Shall we just drop these links?

Hmm, well IMO I really think it would be. For example, for `Type` field, e.g. based on the [PEP Types](https://peps.python.org/pep-0001/#pep-types) descriptions:

* Standards Track: "Normative PEP with a new feature for Python, implementation change for CPython or interoperability standard for the ecosystem"
* Informational: "Non-normative PEP containing background, guidelines or other information relevant to the Python ecosystem"
* Process: "Normative PEP describing or proposing a change to a Python community process, workflow or governance"

And for the `Status` field, based on the [PEP Workflow](https://peps.python.org/pep-0001/#pep-workflow) section (we really should have a bulleted list of the statuses, like we do for Types):

* Draft: "Proposal under active discussion and revision"
* Deferred: "Inactive draft proposal that may be taken up again at a later time"
* Accepted: "Normative proposal accepted for implementation"
* Active: "Currently valid/in-use guidance or process"
* Final: "Accepted and implementation complete, or no longer active"
* Withdrawn: "Removed from consideration by author(s)/sponsor"
* Rejected: "Formally declined and will not be accepted"
* Superseded: "Replaced by another succeeding PEP"

hugovk commented:
Okay, let's try it. Next question is how to create an abbreviation node :)

AA-Turner commented:
`nodes.abbreviation('text', 'text', explanation='blah')` from memory

A

hugovk commented:
Thanks, updated!

Also added status:

* Provisional: "Provisionally accepted but additional feedback needed"

# After

![image](https://user-images.githubusercontent.com/1324225/199026214-4c752b8a-eb50-4df9-83c6-d725fd6a1181.png)

https://pep-previews--2842.org.readthedocs.build/pep-0594/

CAM-Gerlach commented:
Oh, I didn't include provisional since to my understanding it had been semi-retired

CAM-Gerlach commented on pep_sphinx_extensions/pep_processor/transforms/pep_headers.py:
```diff
@@ -223,3 +254,47 @@ def _process_pretty_url(url: str) -> tuple[str, str]:
+    if status in SPECIAL_STATUSES:
+        status = SPECIAL_STATUSES[status]
+
+    if status == STATUS_DRAFT:
+        return "Proposal under active discussion and revision"
+    if status == STATUS_DEFERRED:
+        return "Inactive draft that may be taken up again at a later time"
+    if status == STATUS_ACCEPTED:
+        return "Normative proposal accepted for implementation"
+    if status == STATUS_ACTIVE:
+        return "Currently valid/in-use guidance or process"
+    if status == STATUS_FINAL:
+        return "Accepted and implementation complete, or no longer active"
+    if status == STATUS_WITHDRAWN:
+        return "Removed from consideration by author(s)/sponsor"
+    if status == STATUS_REJECTED:
+        return "Formally declined and will not be accepted"
+    if status == STATUS_SUPERSEDED:
+        return "Replaced by another succeeding PEP"
+    if status == STATUS_PROVISIONAL:
+        return "Provisionally accepted but additional feedback needed"
```
Wouldn't it be nicer and more Python to make this a dictionary lookup table (e.g. in a module-level constant) rather than a long chain of if statements?

hugovk replied:
Yep, refactored!

CAM-Gerlach commented on pep_sphinx_extensions/pep_processor/transforms/pep_headers.py:
```diff
@@ -223,3 +254,47 @@ def _process_pretty_url(url: str) -> tuple[str, str]:
+
+
+def _abbreviate_type(type_: str) -> str:
+    if type_ == "Standards Track":
+        return (
+            "Normative PEP with a new feature for Python, implementation change for "
+            "CPython or interoperability standard for the ecosystem"
+        )
+    if type_ == "Informational":
+        return (
+            "Non-normative PEP containing background, guidelines or other information "
+            "relevant to the Python ecosystem"
+        )
+    if type_ == "Process":
+        return (
+            "Normative PEP describing or proposing a change to a Python community "
+            "process, workflow or governance"
+        )
```
Same here

CAM-Gerlach reviewed (commented):
Seems to work great, but a refactoring comment on the implementation

hugovk commented:
> Oh, I didn't include provisional since to my understanding it had been semi-retired

Okay, shall we remove it from here then, so it would raise an exception if used again?

If so, should we (separately) remove the other handling for it, and update PEP-1?

encukou commented:
>>  Oh, I didn't include provisional since to my understanding it had been semi-retired
>
> Okay, shall we remove it from here then, so it would raise an exception if used again?

In the past, provisional PEPs didn't really work out well, so the current SC informally/internally decided that it's better to not accept PEPs provisionally. So the status probably won't be used any time soon, but it should stay on the list.

CAM-Gerlach commented on pep_sphinx_extensions/pep_zero_generator/writer.py:
```diff
@@ -177,9 +177,10 @@ def write_pep0(self, peps: list[PEP], header: str = HEADER, intro: str = INTRO,
         self.emit_title("PEP Types Key")
         for type_ in sorted(TYPE_VALUES):
-            self.emit_text(f"    {type_[0]} - {type_} PEP")
+            self.emit_text(f"* {type_[0]} - {type_} PEP")
```
```suggestion
            self.emit_text(
                f"* **{type_[0]}** --- *{type_}*: ABBREVIATED_TYPES[type_]")
```

To add the short description of each type to the key, to make it actually useful

Also, needs to import `ABBREVIATED_TYPES` from `pep_headers` (which is possible now that it is a module-level dict constant :) )

CAM-Gerlach commented on pep_sphinx_extensions/pep_zero_generator/writer.py:
```diff
@@ -189,12 +190,13 @@ def write_pep0(self, peps: list[PEP], header: str = HEADER, intro: str = INTRO,
                 continue
             if status == STATUS_ACCEPTED:
-                msg = "    A - Accepted (Standards Track only) or Active proposal"
+                msg = "* A - Accepted (Standards Track only) or Active proposal"
             else:
-                msg = f"    {status[0]} - {status} proposal"
+                msg = f"* {status[0]} - {status} proposal"
```
As mentioned above, all these lines can be removed, as the above line replaces this special casing.

CAM-Gerlach commented on pep_sphinx_extensions/pep_zero_generator/writer.py:
```diff
@@ -189,12 +190,13 @@ def write_pep0(self, peps: list[PEP], header: str = HEADER, intro: str = INTRO,
             else:
-                msg = f"    {status[0]} - {status} proposal"
+                msg = f"* {status[0]} - {status} proposal"
             self.emit_text(msg)
```
```suggestion
            self.emit_text(
                f"* **{status_code}** --- *{status}*: ABBREVIATED_STATUSES[status]")
```

Also, needs to import `ABBREVIATED_STATUSES` from `pep_headers` (which is possible now that it is a module-level dict constant :) )

CAM-Gerlach commented on pep_sphinx_extensions/pep_zero_generator/writer.py:
```diff
@@ -189,12 +190,13 @@ def write_pep0(self, peps: list[PEP], header: str = HEADER, intro: str = INTRO,
             if status in HIDE_STATUS:
                 continue
```
```suggestion
            status_code = "<none>" if status == STATUS_DRAFT else status[0]
```

This whole block of if statements can be replaced with just this line, as Accepted/Active no longer need to be special cased (as their descriptions explain the difference better than the old bespoke text), and Draft PEPs should still be mentioned here somehow, since it doesn't state anywhere in the document that PEPs with an empty status code are draft PEPs.

Also, `HIDE_STATUS` in `pep_zero_generator.constants` can consequently be removed.

CAM-Gerlach reviewed (commented):
The in-PEP abbreviations are LGTM, but I do have a couple comments on the changes to the PEP 0 types/status key, both mostly implemented as suggestions, Specifically:

* [ ] Now that the names of each type/status are shown directly in the tooltips for each PEP, the  top-level "key" sections are redundant, as they just repeat them. However, they could still be useful if you injected the corresponding description for each type and status next to their name. This also avoids special casing Accepted/Active, as their descriptions explain the difference better (if still not perfectly) than what's currently there.

  This looks like:
  * **A** — *Accepted*: Normative proposal accepted for implementation
  * **A** — *Active*: Currently valid/in-use information or process
  etc.
  
  I've added the suggestions to implement this (plus a couple other small changes I couldn't reach, noted there)
  
* [x] There's also nothing I can find anywhere that in PEP 0 that states that Draft PEPs are those with no *type* letter. While it might be nice to add a some character/symbol that could be used for an `:abbr:`, for now I've just made the appropriate suggestion alongside the above such that drafts are at least listed in the key

hugovk commented on pep_sphinx_extensions/pep_zero_generator/writer.py:
```diff
@@ -177,9 +177,10 @@ def write_pep0(self, peps: list[PEP], header: str = HEADER, intro: str = INTRO,
+            self.emit_text(f"* {type_[0]} - {type_} PEP")
             self.emit_newline()
 
+        self.emit_text("More info in :pep:`PEP 1 <1#pep-types>`.")
```
You made three other suggestions, and I chose the first :)

> I'm not sure the section name is needed at all, since you're linking directly to it. You could just say "More info in [PEP 1](https://github.com/python/peps/pull/LINK)" or to clarify both links aren't the same and just to the top level of PEP 1, "See [the descriptions in PEP 1](https://github.com/python/peps/pull/LINK)". Or even just "[More info](https://github.com/python/peps/pull/LINK)" for both.

But happy to go with this one :)

CAM-Gerlach replied:
Oh sorry! Seems like I was the one who got lost, doh!

hugovk commented:
Thanks, updated!

CAM-Gerlach commented on pep_sphinx_extensions/pep_zero_generator/writer.py:
```diff
@@ -177,24 +180,25 @@ def write_pep0(self, peps: list[PEP], header: str = HEADER, intro: str = INTRO,
-            else:
-                msg = f"    {status[0]} - {status} proposal"
-            self.emit_text(msg)
+            status_code = "<none>" if status == STATUS_DRAFT else status[0]
```
We might want to bikeshed a bit over this. E.g. we could use the space _symbol_, which as a single character like the others would make the bullets line up (which the previous throws off)

```suggestion
            status_code = "␣" if status == STATUS_DRAFT else status[0]
```

hugovk replied:
Can't think of a better one, but "<none>" seems better than "␣" 🤷

CAM-Gerlach replied:
I guess so, from a clarity perspective, though it still irks me that it throws the output formatting off...I was originally going to propose using another letter/symbol for Draft (and maybe `Ac`/`At` for Accepted/Active) since we're special casing it anyway, but I didn't want to drag this PR further out of scope than I have already.

Also, fun fact: apparently [`<none>`] displays as [<none>] in GFM. TIL...

AA-Turner replied:
An alternate, more verbose but the alignment is already lost.
```suggestion
            status_code = "No status key" if status == STATUS_DRAFT else status[0]
```

CAM-Gerlach replied:
The "key" here is this _section_, listing the mappings from letters to statuses/types, so saying there is no key in the key is a self-defeating prophecy :) Instead, you could say "No letter" or "No status letter". Also, the `<>` makes clear that it is a description rather than a literal, as the other entries are. So:

```suggestion
            status_code = "<No letter>" if status == STATUS_DRAFT else status[0]
```

or

```suggestion
            status_code = "<No status letter>" if status == STATUS_DRAFT else status[0]
```

CAM-Gerlach reviewed (commented):
One change I can't reach—`HIDE_STATUS` in `constants.py` should be reduced to just Draft, since `Active` is no longer needed there, and furthermore it means Active process PEPs don't have the `A` displayed (as both the old and the new key section state that they should) in either the label or the tooltip, so they are impossible to tell apart from Draft process PEPs (of course, it means also cannot be told apart from Accepted process PEPs by the label, but they can be from the tooltip and at least they are documented as such).

Come to think of it, there's no reason not to include the Draft status in the tooltip even if we don't in the label, to make it clearer to readers. That could be done by simply changing line 134 of `pep_zero_generator/parser.py` from

```python
            return f":abbr:`{type_code} ({self.pep_type})`"
```

to

```python
            return f":abbr:`{type_code} ({self.pep_type}, {self.status})`"
```

hugovk commented:
Thanks, updated!

CAM-Gerlach commented:
You'll need to update a few tests after this latest change (can't reach them myself):

```
FAILED pep_sphinx_extensions/tests/pep_zero_generator/test_parser.py::test_pep_details - AssertionError: assert {'authors': '... Python Code'} == {'authors': '... P...
FAILED pep_sphinx_extensions/tests/pep_zero_generator/test_parser.py::test_abbreviate_type_status[Informational-Draft-:abbr:`I (Informational)`] - AssertionError: assert ':abbr:`I (In...onal, Draft)`' == ':abbr:`I (Informa...
FAILED pep_sphinx_extensions/tests/pep_zero_generator/test_parser.py::test_abbreviate_type_status[Informational-Active-:abbr:`I (Informational)`] - AssertionError: assert ':abbr:`IA (I...nal, Active)`' == ':abbr:`I (Informa...
FAILED pep_sphinx_extensions/tests/pep_zero_generator/test_parser.py::test_abbreviate_type_status[Process-Active-:abbr:`P (Process)`] - AssertionError: assert ':abbr:`PA (Process, Active)`' == ':abbr:`P (Process)`'
```

hugovk commented:
Thanks, fixed!

CAM-Gerlach reviewed (approved):
LGTM for now! Thanks for all your patience and responsiveness after all the iterative feedback.

AA-Turner commented on pep_sphinx_extensions/pep_processor/transforms/pep_headers.py:
```diff
@@ -7,7 +7,42 @@
+    STATUS_DRAFT: "Proposal under active discussion and revision",
+    STATUS_DEFERRED: "Inactive draft that may be taken up again at a later time",
+    STATUS_ACCEPTED: "Normative proposal accepted for implementation",
+    STATUS_ACTIVE: "Currently valid/in-use guidance or process",
```
```suggestion
    STATUS_ACTIVE: "Currently valid or in-use guidance or process",
```

CAM-Gerlach replied:
Well, its not necessarily (exclusive) "or" though; the word "or" at least to me carries an implication of exclusivity and a more tightly defined relationship between the two qualities than a slash, which is deliberately more broad. Furthermore, it results in a double "or", making the _overall grammar of the sentence_ more ambiguous and harder to parse at a glance, whereas the `/` spacing establishes a clearer syntactic hierarchy.

What about the following, which avoids the slash while being easier to read and clearer which adjectives apply to which words, as well as clearer on the distinction between the types of PEPs that Active vs. Accepted generally apply to (to reduce the ambiguity in the key between their identical characters).

```suggestion
    STATUS_ACTIVE: "Currently valid informational guidance, or an in-use process",
```

AA-Turner commented on pep_sphinx_extensions/pep_processor/transforms/pep_headers.py:
```diff
@@ -7,7 +7,42 @@
+    STATUS_ACCEPTED: "Normative proposal accepted for implementation",
+    STATUS_ACTIVE: "Currently valid/in-use guidance or process",
+    STATUS_FINAL: "Accepted and implementation complete, or no longer active",
+    STATUS_WITHDRAWN: "Removed from consideration by author(s)/sponsor",
```
```suggestion
    STATUS_WITHDRAWN: "Removed from consideration by the author or sponsor(s)",
```

CAM-Gerlach replied:
The pluralism was modified here to put the `(s)` on the sponsor, when in fact the sponsor is always singular and it is the authors that may be plural. So, this should read

```suggestion
    STATUS_WITHDRAWN: "Removed from consideration by the sponsor or author(s)",
```

hugovk replied:
Went for a generic "sponsor or authors" to remove the ugly "(s)"

AA-Turner reviewed (commented):
Two notes about avoiding `/` in prose text.

I missed the discussion about Active/Accepted -- I don't think this matters but to flag that on mobile tooltips are nigh on impossible to load (The proportion of our audience browsing the PEP index on mobile is likely limited to just me, so as I say I don't think it should affect the changes here!)

A

AA-Turner commented:
@python/organization-owners please block as spam.

A

hugovk commented:
@AA-Turner Any further comments?

AA-Turner reviewed (approved):
Thanks!

A
